scheduler/system_util_test.go

```diff
@@ -68,205 +85,310 @@ func TestDiffSystemAllocsForNode_Sysbatch_terminal(t *testing.T) {
+			diff := diffSystemAllocsForNode(
+				job, tc.nodeID, eligible, nil,
+				tainted, required, allocsForNode, terminal, true)
+
+			assertDiffCount(t, tc.expected, diff)
+		})
+	}
+}
+
+// TestDiffSystemAllocsForNodes_Stops verifies we stop allocs we no longer need
+func TestDiffSystemAllocsForNode_Stops(t *testing.T) {
+	ci.Parallel(t)
+
+	job := mock.SystemJob()
+	required := materializeSystemTaskGroups(job)
+
+	// The "old" job has a previous modify index but is otherwise unchanged, so
+	// existing non-terminal allocs for this version should be updated in-place
+
+	// TODO(tgross): *unless* there's another alloc for the same job already on
+	// the node. See https://github.com/hashicorp/nomad/pull/16097
+	oldJob := new(structs.Job)
+	*oldJob = *job
+	oldJob.JobModifyIndex -= 1
+
+	node := mock.Node()
+
+	eligible := map[string]*structs.Node{
+		node.ID: node,
 	}
 
 	allocs := []*structs.Allocation{
-		// Update the 1st
 		{
+			// extraneous alloc for old version of job should be updated
+			// TODO(tgross): this should actually be stopped.
+			// See https://github.com/hashicorp/nomad/pull/16097
 			ID:     uuid.Generate(),
-			NodeID: "zip",
+			NodeID: node.ID,
 			Name:   "my-job.web[0]",
 			Job:    oldJob,
 		},
-
-		// Ignore the 2rd
-		{
+		{ // most recent alloc for current version of job should be ignored
 			ID:     uuid.Generate(),
-			NodeID: "zip",
-			Name:   "my-job.web[1]",
+			NodeID: node.ID,
+			Name:   "my-job.web[0]",
 			Job:    job,
 		},
-
-		// Evict 11th
-		{
+		{ // task group not required, should be stopped
 			ID:     uuid.Generate(),
-			NodeID: "zip",
-			Name:   "my-job.web[10]",
-			Job:    oldJob,
+			NodeID: node.ID,
+			Name:   "my-job.something-else[0]",
+			Job:    job,
 		},
+	}
 
-		// Migrate the 3rd
-		{
-			ID:     uuid.Generate(),
-			NodeID: "drainNode",
-			Name:   "my-job.web[2]",
-			Job:    oldJob,
-			DesiredTransition: structs.DesiredTransition{
-				Migrate: pointer.Of(true),
+	tainted := map[string]*structs.Node{}
+	terminal := structs.TerminalByNodeByName{}
+
+	diff := diffSystemAllocsForNode(
+		job, node.ID, eligible, nil, tainted, required, allocs, terminal, true)
+
+	assertDiffCount(t, diffResultCount{ignore: 1, stop: 1, update: 1}, diff)
+	if len(diff.update) > 0 {
+		test.Eq(t, allocs[0], diff.update[0].Alloc)
+	}
+	if len(diff.ignore) > 0 {
+		test.Eq(t, allocs[1], diff.ignore[0].Alloc)
+	}
+	if len(diff.stop) > 0 {
+		test.Eq(t, allocs[2], diff.stop[0].Alloc)
+	}
+}
+
+// Test the desired diff for an updated system job running on a ineligible node
+func TestDiffSystemAllocsForNode_IneligibleNode(t *testing.T) {
+	ci.Parallel(t)
+
+	job := mock.SystemJob()
+	required := materializeSystemTaskGroups(job)
+
+	ineligibleNode := mock.Node()
+	ineligibleNode.SchedulingEligibility = structs.NodeSchedulingIneligible
+	ineligible := map[string]struct{}{
+		ineligibleNode.ID: {},
+	}
+
+	eligible := map[string]*structs.Node{}
+	tainted := map[string]*structs.Node{}
```

Comment:
`ineligible` is the node has been marked `ineligible` for scheduling by the user (usually but not necessarily as part of a drain) and can't get any _new_ placements but we leave the existing allocs alone (unless the node is `ineligible` because it's been marked for draining). `tainted` means this node is down.